File: csharp/nucleotide-count/NucleotideCount.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class NucleotideCount
{
    public static IDictionary<char, int> Count(string sequence)
    {
        var map = new Dictionary<char, int> { { 'A', 0 }, { 'G', 0 }, { 'C', 0 }, { 'T', 0 } };
        return sequence.Aggregate(map, (acc, c) =>
        {
            if (!acc.ContainsKey(c)) { throw new ArgumentException("invalid char"); }
            ++acc[c];
            return acc;
        });
    }
}

File: cpp/nucleotide-count/nucleotide_count.cpp
#include "nucleotide_count.h"

using namespace std;

map<char, int> nucleotide_count::count(const string &s) {
    map<char, int> dict({
                                {'A', 0},
                                {'T', 0},
                                {'C', 0},
                                {'G', 0}
                        });
    for (const char &c: s) {
        if (dict.find(c) == dict.end()) {
            throw invalid_argument(string("invalid nucleotide: ") + c);
        }
        ++dict[c];

    }

    return dict;
}
